common/metrics/tags.go

```diff
@@ -58,14 +62,23 @@ type (
 	}
 )
 
+// sanitizer is used to sanitize tag values, to prevent collisions and invalid data.
+// values should not allow dots for "namespacing" / "grouping" metrics, as they seem to refer to "one thing" in all cases.
+var sanitizer = tally.NewSanitizer(tally.SanitizeOptions{
+	NameCharacters:       tally.ValidCharacters{tally.AlphanumericRange, tally.UnderscoreDashCharacters},
+	KeyCharacters:        tally.ValidCharacters{tally.AlphanumericRange, tally.UnderscoreDashCharacters},
+	ValueCharacters:      tally.ValidCharacters{tally.AlphanumericRange, tally.UnderscoreDashCharacters},
+	ReplacementCharacter: '_',
+})
+
 func (s simpleMetric) Key() string   { return s.key }
 func (s simpleMetric) Value() string { return s.value }
 
 func metricWithUnknown(key, value string) Tag {
 	if len(value) == 0 {
 		value = unknownValue
 	}
-	return simpleMetric{key: key, value: value}
+	return simpleMetric{key: key, value: sanitizer.Value(value)}
```

Comment:
I was looking into the same issue as some metrics with value constructed by our java client was dropped by the metrics system due to invalid characters. I also sent out an announcement in our open source slack channel to see if this change will impact open source customer or not.

Will doing sanitization for all metrics tag too expensive and cause performance issues? My original thought is that we can only apply sanitizer to the values we don't have control over, for example, tasklist, workflow type, activity type. For other types the value is set by the server code so we know their value won't contain invalid characters.

Reply:
sanitization cost: should be practically non-existent, it's doing a very simple transformation on a fairly short string.  other code in metrics alone likely costs *significantly* more, to say nothing of all the other code being executed.